extensions/ql-vscode/src/queryRunner.ts

```diff
@@ -54,4 +97,53 @@ export abstract class QueryRunner {
   ): Promise<void>;
 
   abstract clearPackCache(): Promise<void>;
+
+  /**
+   * Create a `CoreQueryRun` object. This creates an object whose `evaluate()` function can be
+   * called to actually evaluate the query. The returned object also contains information about the
+   * query evaluation that is known even before evaluation starts, including the unique ID of the
+   * evaluation and the path to its output directory.
+   */
+  public createQueryRun(
+    dbPath: string,
+    query: CoreQueryTarget,
+    generateEvalLog: boolean,
+    additionalPacks: string[],
+    queryStorageDir: string,
+    id: string | undefined,
+    templates: Record<string, string> | undefined,
+  ): CoreQueryRun {
+    const actualId = id ?? `${basename(query.queryPath)}-${nanoid()}`;
+    const outputDir = new QueryOutputDir(join(queryStorageDir, actualId));
+
+    return {
+      queryTarget: query,
+      dbPath,
+      id: actualId,
+      outputDir,
+      evaluate: async (
+        progress: ProgressCallback,
+        token: CancellationToken,
+        logger: BaseLogger,
+      ): Promise<CoreCompletedQuery> => {
```

Comment:
`evaluate()` needs to use several of the parameters already passed to `createQueryRun()`, so one way or another `createQueryRun()` has to return a new object to hold that state. I could make `CoreQueryRun` into a class, and make the state capture explicit. Would that make things more clear overall? Or are you suggesting that `evaluate()` be a member of `QueryRunner` but take a `CoreQueryRun` as a parameter?

Reply:
I also was going to make a comment asking whether the `evaluate` method needs to live on this interface. It's a bit of a shame that we then have to explicitly omit it when defining the `CoreCompletedQuery` interface.

I note there's only one implementation, but I see what you mean about it needing access to state available in `createQueryRun`.  But in the two places where we call `evaluate`, we also called `createQueryRun` earlier in that same method, so the state would still be there and could be passed into `evaluate` as well / instead.

Reply:
While it's not apparent in this PR, my motivation for having `CoreQueryRun` as an object with an `evaluate()` function, rather than having two different functions on `QueryRunner` that require several common parameters, was that the debug adapter needs to initialize an evaluation in one function, but then later await it in another function. A `CoreQueryRun` object seemed like the natural way to encapsulate everything the debug adapter would need to later evaluate the query. I'd rather not have to have the debugger capture multiple parameters from the `createQueryRun()` method just to pass them to a later `evaluateQueryRun()` method.

The real problem here is that I want a function that evaluates a query, which is asynchronous, and so would normally return a `Promise<CoreQueryResults>`. However, just _starting_ the evaluation produces some properties, like `outputDir`, that I'll need to use before the promise completes. I could just have a single `runQuery()` function on `QueryRunner`, and have it return something like `Promise<CoreQueryResult> & CoreQueryRun` (note that `CoreQueryRun` is not inside the angle brackets). I'm pretty sure I've seen this done somewhere before. I'm not sure if other developers would find it more or less odd than the separate `evaluate()` function I have now.

I suppose I could just replace the `evaluate` property in `CoreQueryRun` with `results: Promise<CoreQueryResults>`. That seems less confusing than anything I suggest above, and also less confusing than what I originally wrote.